File: NBlast.Api/Async/IndexingQueueKeeper.fs
namespace NBlast.Api.Async

open NBlast.Api.Models
open NBlast.Storage
open NBlast.Storage.Core.Index
open NBlast.Storage.Core.Extensions
open System.Collections.Concurrent
open System.Threading.Tasks

type IQueueKeeper<'a> = interface
    abstract member Enqueue : 'a -> Task<unit>
    abstract member Consume : unit -> 'a option
    abstract member ConsumeMany : int option -> seq<'a>
    abstract member Count : unit -> int 
end

type IndexingQueueKeeper() =
    static let logger = NLog.LogManager.GetCurrentClassLogger()
    let queue = new ConcurrentQueue<LogModel>()

    interface IQueueKeeper<LogModel> with
        member me.Count() = queue |> Seq.length 
        member me.Enqueue model = 
            async {
                model |> sprintf "Enqueuing model %A" |> logger.Debug
                queue.Enqueue(model)
            } 
            |> Async.StartAsTask

        member me.Consume() = 
            match queue.TryDequeue() with
            | (false, _) -> None
            | (true, model) ->
                model |> sprintf "Dequeuing model %A" |> logger.Debug 
                Some model

        member me.ConsumeMany ?amount =
            let consume = (me :> IQueueKeeper<LogModel>).Consume
            [ 0 .. (amount |? 10) - 1 ] |> Seq.map (fun x -> consume()) 
                |> Seq.filter Option.isSome 
                |> Seq.map Option.get 
            

File: NBlast.Rest/Async/IndexingQueueKeeper.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using LanguageExt;
using NBlast.Rest.Model.Write;
using Serilog;
using static LanguageExt.Prelude;

namespace NBlast.Rest.Async
{
    public class IndexingQueueKeeper: IIndexingQueueKeeper
    {
        private static readonly ILogger Logger = Log.Logger.ForContext<IndexingQueueKeeper>();
        private readonly ConcurrentQueue<LogEvent> _queue = new ConcurrentQueue<LogEvent>();

        public Option<LogEvent> Consume()
        {
            LogEvent result;
            return _queue.TryDequeue(out result) ? Some(result) : None;
        }

        public Option<LogEvent> Peek()
        {
            LogEvent result;
            return _queue.TryPeek(out result) ? Some(result) : None;
        }

        public int Count() => _queue.Count;

        public IReadOnlyList<LogEvent> ConsumeTop(int top = 10) =>
            Range(0, top - 1)
                .Select(x => Consume())
                .Select(x => x.Match(Some: y => y,
                                     None: () => null))
                .Where(x => x != null)
                .ToImmutableList();
        

        public IReadOnlyList<LogEvent> PeekTop(int top = 10) => _queue.Take(top).ToImmutableList();

        public Unit Enqueue(LogEvent entry)
        {
            Logger.Debug("Enqueuing model @{0}", entry);
            _queue.Enqueue(entry);
            return unit;
        }
    }
}